core/shared/src/main/scala/zio/logging/LoggerNameExtractor.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2019-2022 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zio.logging
+
+import zio.{ FiberRefs, Trace }
+
+trait LoggerNameExtractor { self =>
+
+  def apply(
+    trace: Trace,
+    context: FiberRefs,
+    annotations: Map[String, String]
+  ): String
```

Comment:
If you change this `String` to `Option[String]`, then you can define `||` (or else) on `LoggerNameExtractor`, which will let you implement `annotationOrTrace` as `annotation || trace` (a more compositional definition).